File: implementations/typescript/packages/llmproxy/src/refresh.ts
// SPDX-License-Identifier: Apache-2.0
import {canonicalVersion,type Envelope} from "@psp-cdl/core";
import {SecurityService,ServiceError,type Principal} from "@psp-cdl/api-server";
import {WorkflowStore,StoreError,bounded,integer,validPromptState,comparePromptVersions,PROMPT_REFRESH_PROFILE,type OwnerReservation} from "@psp-cdl/api-server/persistence";
import {McpDispatchGate,bindingDigest} from "@psp-cdl/mcpproxy";
import {BufferedLlmLoop,LoopError,type LoopHost,type LoopOptions,type ProviderRegistration} from "./loop.js";
import {DurableLlmLoop,type DurableHost,type DurableOptions} from "./durable.js";
export {PROMPT_REFRESH_PROFILE,comparePromptVersions};
const attributes=["refresh-policy","refresh-interval","refresh-grace"];
const fail=(code:string):never=>{throw new LoopError(code);};
const copy=(v:unknown):any=>{try{return bounded(v);}catch{return fail("INVALID_REFRESH");}};
const owner=(p:Principal)=>({tenantId:p.tenantId,subjectId:p.subjectId});
async function call<T>(f:()=>T|Promise<T>):Promise<T>{try{return await f();}catch{return fail("HOST_ERROR");}}
type State=Record<string,any>;
export interface RefreshHost extends DurableHost {
  refresh(principal:Principal,binding:Record<string,unknown>,request:Record<string,unknown>):Envelope|Promise<Envelope>;
  authorizeRefresh(principal:Principal,binding:Record<string,unknown>,previous:Record<string,unknown>|null,candidate:Envelope):boolean|Promise<boolean>;
  auditRefresh(principal:Principal,event:Record<string,unknown>):boolean|Promise<boolean>;
}
function describe(prompt:Envelope):State {
  const s=prompt.signature,a=s.attributes??{},policy=a["refresh-policy"]??"expiration",policies=policy.split('|').sort();
  if(!policies.length||policies.some((p,i)=>!["expiration","interval"].includes(p)||policies.indexOf(p)!==i))fail("UNSUPPORTED_REFRESH");
  const decimal=(v:string|undefined,defaultValue:number)=>{if(v===undefined)return defaultValue;if(!/^(0|[1-9][0-9]*)$/.test(v)||/[^0-9]/.test(v)||!integer(Number(v)))return fail("INVALID_REFRESH");return Number(v);};
  const interval=decimal(a["refresh-interval"],0),grace=decimal(a["refresh-grace"],300);
  if(policies.includes("interval")!==(interval>0)||!policies.includes("interval")&&a["refresh-interval"]!==undefined)fail("INVALID_REFRESH");
  const text=(prompt.data as string).replace(/\r\n?/g,'\n').replace(/^[\t\n ]+|[\t\n ]+$/g,'');
  return {version:canonicalVersion(s.version),digest:bindingDigest({text,trustLevel:s.trustLevel??2,priority:s.priority??50,policies,interval,grace}),policies,interval,grace,timestamp:s.timestamp,expires:s.expires,turnCount:0,refreshCount:0};
}
async function saved(store:WorkflowStore,p:Principal,b:Record<string,unknown>):Promise<State|null> {
  try {const r=await store.execute(owner(p),{action:"getPromptState",sessionId:b.sessionId});if(!validPromptState(r.state))fail("INVALID_REFRESH_STATE");return r;}
  catch(e){if(e instanceof StoreError&&e.code==="NOT_FOUND")return null;throw e;}
}
async function pinned(store:WorkflowStore,p:Principal,b:Record<string,unknown>,prompt:Envelope):Promise<void> {
  const r=await saved(store,p,b),d=describe(prompt);
  if(!r||r.state.version!==d.version||r.state.digest!==d.digest)fail("STALE_PROMPT");
}
class RefreshBuffer extends BufferedLlmLoop {
  revision=0;
  private current:State|null=null;
  constructor(store:WorkflowStore,gate:McpDispatchGate,host:LoopHost,provider:ProviderRegistration,
    private refreshHost:RefreshHost,private onPrompt:(p:Envelope)=>void,private token:unknown,private options:DurableOptions){super(store,gate,host,provider);}
  protected override promptAttributes():string[]{return attributes;}
  protected override async verify(p:Principal,b:Record<string,unknown>,prompt:Envelope):Promise<void>{await super.verify(p,b,prompt);await pinned(this.store,p,b,prompt);}
  private async ensure(p:Principal,b:Record<string,unknown>):Promise<void>{
    this.check(this.options);
    const live=await new SecurityService({authenticate:t=>this.host.authenticate(t),now:()=>this.host.now(),resolve:()=>null}).authenticate(this.token);
    if(live.tenantId!==p.tenantId||live.subjectId!==p.subjectId||!["models:invoke","sessions:write"].every(s=>live.scopes.includes(s)))fail("FORBIDDEN");
    const s=await this.store.execute(owner(p),{action:"getSession",sessionId:b.sessionId});
    if(s.version!==b.sessionVersion||s.status!=="running")fail("STALE_SESSION");
    const a=await this.snapshot(live,s);this.check(this.options,Math.min(a.expires,s.expiresAt as number));
    if(a.revision!==b.authorityRevision||a.policyVersion!==b.policyVersion)fail("STALE_AUTHORITY");
  }
  private due(state:State,now:number):string|null {
    if(now>=state.expires||state.policies.includes("expiration")&&state.expires-now<=state.grace)return "expiration";
    if(state.policies.includes("interval")&&state.turnCount>=state.interval)return "interval";
    return null;
  }
  private async audit(p:Principal,b:Record<string,unknown>,signal:string,details:State):Promise<void>{
    let ok;try{ok=await this.refreshHost.auditRefresh(copy(p),copy({profile:PROMPT_REFRESH_PROFILE,signal,binding:b,at:this.host.now(),...details}));}catch{fail("AUDIT_FAILED");}
    if(ok!==true)fail("AUDIT_FAILED");
  }
  private async install(p:Principal,b:Record<string,unknown>,reservation:OwnerReservation,trigger:string|null):Promise<Envelope>{
    const previous=this.current?.state??null,started=this.host.now();
    try {
      await this.ensure(p,b);
      const request={session_id:b.sessionId,current_version:previous?.version??null,trigger:trigger??"initial",turn_count:previous?.turnCount??0};
      const prompt=copy(await call(()=>trigger?this.refreshHost.refresh(copy(p),copy(b),copy(request)):this.host.prompt(copy(p),copy({...b,refresh:request})))) as Envelope;
      await super.verify(p,b,prompt);const next=describe(prompt);
      if(previous){
        const comparison=comparePromptVersions(next.version,previous.version);
        if(comparison<0)fail("PROMPT_ROLLBACK");
        if(comparison===0&&next.digest!==previous.digest)fail("PROMPT_VERSION_CONFLICT");
        if(next.timestamp<started||next.timestamp<=previous.timestamp||next.expires-this.host.now()<=next.grace)fail("REFRESH_NOT_FRESH");
        next.refreshCount=previous.refreshCount+1;
      }
      if(await call(()=>this.refreshHost.authorizeRefresh(copy(p),copy({...b,trigger:trigger??"initial"}),copy(previous),copy(prompt)))!==true)fail("REFRESH_DENIED");
      await this.ensure(p,b);await super.verify(p,b,prompt);
      let guardError:unknown;
      const result=await this.store.execute(owner(p),{action:"putPromptState",sessionId:b.sessionId,expectedVersion:b.sessionVersion,refreshRevision:this.revision,state:next},async()=>{
        try{await this.ensure(p,b);await super.verify(p,b,prompt);return true;}catch(e){guardError=e;return false;}
      },reservation).catch(e=>{throw guardError??e;});
      this.current=result;this.revision=result.revision as number;
      await this.audit(p,b,trigger?"prompt_refreshed":"prompt_initialized",{trigger:trigger??"initial",previousVersion:previous?.version??null,version:next.version,digest:next.digest});
      await this.ensure(p,b);await this.verify(p,b,prompt);this.onPrompt(prompt);return prompt;
    }catch(e){
      const code=e instanceof LoopError?e.code:e instanceof ServiceError&&["UNAUTHENTICATED","FORBIDDEN"].includes(e.code)?e.code:e instanceof StoreError?["STATE_CONFLICT","PERSISTENCE_DENIED"].includes(e.code)?e.code:"HOST_ERROR":"HOST_ERROR";
      if(code!=="AUDIT_FAILED")await this.audit(p,b,code==="PROMPT_ROLLBACK"?"prompt_refresh_rollback":"prompt_refresh_failed",{trigger:trigger??"initial",previousVersion:previous?.version??null,code});
      return fail(code);
    }
  }
  protected override async loadPrompt(p:Principal,b:Record<string,unknown>,_options:LoopOptions,reservation:OwnerReservation):Promise<Envelope>{
    this.current=await saved(this.store,p,b);this.revision=this.current?.revision??0;
    if(this.current&&this.current.sessionVersion!==b.sessionVersion)fail("STALE_PROMPT");
    const trigger=this.current?this.due(this.current.state,this.host.now()):null;
    if(!this.current||trigger)return this.install(p,b,reservation,trigger);
    const prompt=copy(await call(()=>this.host.prompt(copy(p),copy({...b,refresh:{current_version:this.current!.state.version,trigger:"binding",turn_count:this.current!.state.turnCount}})))) as Envelope;
    await this.verify(p,b,prompt);
    await this.audit(p,b,"prompt_bound",{trigger:"binding",previousVersion:this.current!.state.version,version:this.current!.state.version,digest:this.current!.state.digest});
    await this.ensure(p,b);await this.verify(p,b,prompt);this.onPrompt(prompt);return prompt;
  }
  protected override async inferencePrompt(p:Principal,b:Record<string,unknown>,prompt:Envelope,_options:LoopOptions,reservation:OwnerReservation):Promise<Envelope>{
    const current=await saved(this.store,p,b);
    if(!current||current.revision!==this.revision||current.sessionVersion!==b.sessionVersion)fail("STALE_PROMPT");
    const trigger=this.due({...current!.state,expires:Math.min(current!.state.expires,prompt.signature.expires)},this.host.now());
    return trigger?this.install(p,b,reservation,trigger):prompt;
  }
}
/** Boundary-triggered refresh; the signing/compatibility authority is exclusively host-owned. */
export class RefreshingLlmLoop extends DurableLlmLoop {
  constructor(store:WorkflowStore,gate:McpDispatchGate,private refreshHost:RefreshHost,provider:ProviderRegistration,configuration:{postCompletion:"lockdown"}){
    super(store,gate,refreshHost,provider,configuration);
    if(!store.promptRefresh||[refreshHost.refresh,refreshHost.authorizeRefresh,refreshHost.auditRefresh].some(f=>typeof f!=="function"))fail("INVALID_CONFIGURATION");
  }
  protected override promptAttributes():string[]{return attributes;}
  protected override async verify(p:Principal,b:Record<string,unknown>,prompt:Envelope):Promise<void>{await super.verify(p,b,prompt);await pinned(this.store,p,b,prompt);}
  protected override makeBuffered(host:LoopHost,onPrompt:(prompt:Envelope)=>void,token:unknown,_session:Record<string,unknown>,options:DurableOptions):BufferedLlmLoop {
    return new RefreshBuffer(this.store,this.gate,host,this.provider,this.refreshHost,onPrompt,token,options);
  }
  protected override turnCommand(command:Record<string,unknown>,buffered:BufferedLlmLoop):Record<string,unknown>{return {...command,action:"commitRefreshedTurn",refreshRevision:(buffered as RefreshBuffer).revision};}
}

File: implementations/python/packages/llmproxy/src/psp_cdl_llmproxy/refresh.py
# SPDX-License-Identifier: Apache-2.0
"""Boundary-triggered refresh; signing and compatibility remain host authorities."""
import re
from psp_cdl_core import canonical_version
from psp_cdl_api_server.service import SecurityService, ServiceError
from psp_cdl_api_server.persistence import StoreError, integer, valid_prompt_state, compare_prompt_versions, PROMPT_REFRESH_PROFILE
from psp_cdl_mcpproxy import binding_digest
from .loop import BufferedLlmLoop, LoopError, copy
from .durable import DurableLlmLoop, call, actor

ATTRIBUTES=["refresh-policy","refresh-interval","refresh-grace"]

def describe(prompt):
    s=prompt["signature"];a=s.get("attributes",{})
    policies=sorted(a.get("refresh-policy","expiration").split("|"))
    if not policies or any(p not in ("expiration","interval") or policies.index(p)!=i for i,p in enumerate(policies)): raise LoopError("UNSUPPORTED_REFRESH")
    def decimal(value,default):
        if value is None:return default
        if not re.fullmatch(r"0|[1-9][0-9]*",value) or len(value)>16:raise LoopError("INVALID_REFRESH")
        n=int(value)
        if not integer(n):raise LoopError("INVALID_REFRESH")
        return n
    interval=decimal(a.get("refresh-interval"),0);grace=decimal(a.get("refresh-grace"),300)
    if ("interval" in policies)!=(interval>0) or "interval" not in policies and "refresh-interval" in a:raise LoopError("INVALID_REFRESH")
    text=prompt["data"].replace("\r\n","\n").replace("\r","\n").strip("\t\n ")
    return {"version":canonical_version(s["version"]),"digest":binding_digest({"text":text,"trustLevel":s.get("trustLevel",2),"priority":s.get("priority",50),"policies":policies,"interval":interval,"grace":grace}),"policies":policies,"interval":interval,"grace":grace,"timestamp":s["timestamp"],"expires":s["expires"],"turnCount":0,"refreshCount":0}

def saved(store,p,b):
    try:
        result=store.execute(actor(p),{"action":"getPromptState","sessionId":b["sessionId"]})
        if not valid_prompt_state(result["state"]):raise LoopError("INVALID_REFRESH_STATE")
        return result
    except StoreError as exc:
        if exc.code=="NOT_FOUND":return None
        raise

def pinned(store,p,b,prompt):
    r=saved(store,p,b);d=describe(prompt)
    if r is None or r["state"]["version"]!=d["version"] or r["state"]["digest"]!=d["digest"]:raise LoopError("STALE_PROMPT")

class RefreshBuffer(BufferedLlmLoop):
    def __init__(self,store,gate,host,provider,refresh_host,on_prompt,token,options):
        super().__init__(store,gate,host,provider)
        self.refresh_host,self.on_prompt,self.token,self.options=refresh_host,on_prompt,token,options
        self.current,self.revision=None,0
    def _prompt_attributes(self):return ATTRIBUTES
    def _verify(self,p,b,prompt):
        super()._verify(p,b,prompt)
        pinned(self._store,p,b,prompt)
    def ensure(self,p,b):
        self._check(self.options)
        live=SecurityService(self._host).authenticate(self.token)
        if any(live[k]!=p[k] for k in ("tenantId","subjectId")) or any(s not in live["scopes"] for s in ("models:invoke","sessions:write")):raise LoopError("FORBIDDEN")
        s=self._store.execute(actor(p),{"action":"getSession","sessionId":b["sessionId"]})
        if s["version"]!=b["sessionVersion"] or s["status"]!="running":raise LoopError("STALE_SESSION")
        a=self._snapshot(live,s);self._check(self.options,min(a["expires"],s["expiresAt"]))
        if a["revision"]!=b["authorityRevision"] or a["policyVersion"]!=b["policyVersion"]:raise LoopError("STALE_AUTHORITY")
    @staticmethod
    def due(state,now):
        if now>=state["expires"] or "expiration" in state["policies"] and state["expires"]-now<=state["grace"]:return "expiration"
        if "interval" in state["policies"] and state["turnCount"]>=state["interval"]:return "interval"
        return None
    def audit(self,p,b,signal,details):
        try:ok=self.refresh_host.audit_refresh(copy(p),copy({"profile":PROMPT_REFRESH_PROFILE,"signal":signal,"binding":b,"at":self._host.now(),**details}))
        except Exception:raise LoopError("AUDIT_FAILED") from None
        if ok is not True:raise LoopError("AUDIT_FAILED")
    def install(self,p,b,reservation,trigger):
        previous=self.current["state"] if self.current else None
        started=self._host.now()
        try:
            self.ensure(p,b)
            request={"session_id":b["sessionId"],"current_version":previous["version"] if previous else None,"trigger":trigger or "initial","turn_count":previous["turnCount"] if previous else 0}
            prompt=copy(call(lambda:self.refresh_host.refresh(copy(p),copy(b),copy(request)) if trigger else self._host.prompt(copy(p),copy({**b,"refresh":request}))),"INVALID_REFRESH")
            super()._verify(p,b,prompt);next_state=describe(prompt)
            if previous:
                comparison=compare_prompt_versions(next_state["version"],previous["version"])
                if comparison<0:raise LoopError("PROMPT_ROLLBACK")
                if comparison==0 and next_state["digest"]!=previous["digest"]:raise LoopError("PROMPT_VERSION_CONFLICT")
                if next_state["timestamp"]<started or next_state["timestamp"]<=previous["timestamp"] or next_state["expires"]-self._host.now()<=next_state["grace"]:raise LoopError("REFRESH_NOT_FRESH")
                next_state["refreshCount"]=previous["refreshCount"]+1
            if call(lambda:self.refresh_host.authorize_refresh(copy(p),copy({**b,"trigger":trigger or "initial"}),copy(previous),copy(prompt))) is not True:raise LoopError("REFRESH_DENIED")
            self.ensure(p,b);super()._verify(p,b,prompt)
            guard_error=[]
            def guard(_):
                try:
                    self.ensure(p,b);super(RefreshBuffer,self)._verify(p,b,prompt)
                    return True
                except Exception as exc:
                    guard_error.append(exc)
                    return False
            try:result=self._store.execute(actor(p),{"action":"putPromptState","sessionId":b["sessionId"],"expectedVersion":b["sessionVersion"],"refreshRevision":self.revision,"state":next_state},guard,reservation)
            except Exception:
                if guard_error:raise guard_error[0]
                raise
            self.current,self.revision=result,result["revision"]
            self.audit(p,b,"prompt_refreshed" if trigger else "prompt_initialized",{"trigger":trigger or "initial","previousVersion":previous["version"] if previous else None,"version":next_state["version"],"digest":next_state["digest"]})
            self.ensure(p,b);self._verify(p,b,prompt);self.on_prompt(prompt)
            return prompt
        except Exception as exc:
            code=exc.code if isinstance(exc,LoopError) or isinstance(exc,StoreError) and exc.code in ("STATE_CONFLICT","PERSISTENCE_DENIED") or isinstance(exc,ServiceError) and exc.code in ("UNAUTHENTICATED","FORBIDDEN") else "HOST_ERROR"
            if code!="AUDIT_FAILED":self.audit(p,b,"prompt_refresh_rollback" if code=="PROMPT_ROLLBACK" else "prompt_refresh_failed",{"trigger":trigger or "initial","previousVersion":previous["version"] if previous else None,"code":code})
            raise LoopError(code) from None
    def _load_prompt(self,p,b,options,reservation):
        self.current=saved(self._store,p,b);self.revision=self.current["revision"] if self.current else 0
        if self.current and self.current["sessionVersion"]!=b["sessionVersion"]:raise LoopError("STALE_PROMPT")
        trigger=self.due(self.current["state"],self._host.now()) if self.current else None
        if self.current is None or trigger:return self.install(p,b,reservation,trigger)
        prompt=copy(call(lambda:self._host.prompt(copy(p),copy({**b,"refresh":{"current_version":self.current["state"]["version"],"trigger":"binding","turn_count":self.current["state"]["turnCount"]}}))),"INVALID_REFRESH")
        self._verify(p,b,prompt)
        self.audit(p,b,"prompt_bound",{"trigger":"binding","previousVersion":self.current["state"]["version"],"version":self.current["state"]["version"],"digest":self.current["state"]["digest"]})
        self.ensure(p,b);self._verify(p,b,prompt);self.on_prompt(prompt)
        return prompt
    def _inference_prompt(self,p,b,prompt,options,reservation):
        current=saved(self._store,p,b)
        if current is None or current["revision"]!=self.revision or current["sessionVersion"]!=b["sessionVersion"]:raise LoopError("STALE_PROMPT")
        trigger=self.due({**current["state"],"expires":min(current["state"]["expires"],prompt["signature"]["expires"])},self._host.now())
        return self.install(p,b,reservation,trigger) if trigger else prompt

class RefreshingLlmLoop(DurableLlmLoop):
    def __init__(self,store,gate,host,provider,configuration=None):
        super().__init__(store,gate,host,provider,configuration)
        if not store.prompt_refresh or any(not callable(getattr(host,k,None)) for k in ("refresh","authorize_refresh","audit_refresh")):raise LoopError("INVALID_CONFIGURATION")
    def _prompt_attributes(self):return ATTRIBUTES
    def _verify(self,p,b,prompt):
        super()._verify(p,b,prompt)
        pinned(self._store,p,b,prompt)
    def _make_buffered(self,host,on_prompt,token,session,options):return RefreshBuffer(self._store,self._gate,host,self._provider,self._host,on_prompt,token,options)
    def _turn_command(self,command,buffered):return {**command,"action":"commitRefreshedTurn","refreshRevision":buffered.revision}
